File: _scripts/fetch_reviews.py
#!/usr/bin/env python3
"""
Fetches Google reviews via Places API (New) and merges with manually curated reviews.
Updates only the 'google' section of reviews.json — manual/Nextdoor entries preserved.
"""
import json, os, urllib.request, urllib.error, sys

API_KEY  = os.environ.get('GOOGLE_API_KEY', '')
PLACE_ID = 'ChIJx8wLC4K11wwRbfe7hhZiHXs'

if not API_KEY:
    print("ERROR: GOOGLE_API_KEY environment variable not set.")
    sys.exit(1)

url = f'https://places.googleapis.com/v1/places/{PLACE_ID}'
fields = 'id,displayName,rating,userRatingCount,reviews'

req = urllib.request.Request(
    url,
    headers={
        'X-Goog-Api-Key': API_KEY,
        'X-Goog-FieldMask': fields,
    },
)

print(f"Fetching reviews for Place ID {PLACE_ID}...")
try:
    with urllib.request.urlopen(req) as resp:
        data = json.loads(resp.read())
except urllib.error.HTTPError as e:
    body = e.read().decode()
    print(f"HTTP {e.code}: {body}")
    sys.exit(1)

reviews = []
for r in data.get('reviews', []):
    reviews.append({
        'author': r.get('authorAttribution', {}).get('displayName', ''),
        'rating': r.get('rating', 5),
        'text':   r.get('text', {}).get('text', ''),
        'time':   r.get('relativePublishTimeDescription', ''),
        'source': 'google',
    })

google_block = {
    'place_id': data.get('id', PLACE_ID),
    'rating':   data.get('rating', 0),
    'total':    data.get('userRatingCount', 0),
    'reviews':  reviews,
}

reviews_path = os.path.join(os.path.dirname(__file__), '..', 'reviews.json')
existing = {}
if os.path.exists(reviews_path):
    with open(reviews_path) as f:
        existing = json.load(f)

existing['google'] = google_block

with open(reviews_path, 'w') as f:
    json.dump(existing, f, indent=2)

print(f"Done. {len(reviews)} reviews fetched. Rating: {google_block['rating']} ({google_block['total']} total)")

File: astro/scripts/fetch-reviews.mjs
// Refresh src/data/reviews.json from Google Places API (New).
// Runs as `prebuild`; silently keeps the committed JSON if the key is missing or the call fails.
import { readFileSync, writeFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const here = dirname(fileURLToPath(import.meta.url));
const target = join(here, '..', 'src', 'data', 'reviews.json');
const PLACE_ID = 'ChIJx8wLC4K11wwRbfe7hhZiHXs';

function loadKey() {
  if (process.env.PLACES_SERVER_KEY) return process.env.PLACES_SERVER_KEY;
  for (const p of [join(here, '..', '.env'), join(here, '..', '..', 'app', '.env')]) {
    try {
      const m = readFileSync(p, 'utf8').match(/^PLACES_SERVER_KEY=(.+)$/m);
      if (m) return m[1].trim();
    } catch {}
  }
  return null;
}

const key = loadKey();
if (!key) { console.log('[reviews] PLACES_SERVER_KEY not set; keeping committed reviews.json'); process.exit(0); }

try {
  const res = await fetch(`https://places.googleapis.com/v1/places/${PLACE_ID}`, {
    headers: { 'X-Goog-Api-Key': key, 'X-Goog-FieldMask': 'rating,userRatingCount,reviews' },
  });
  const d = await res.json();
  if (!res.ok || !d.userRatingCount) throw new Error(JSON.stringify(d).slice(0, 300));
  const current = JSON.parse(readFileSync(target, 'utf8'));
  const reviews = (d.reviews || []).map((r) => ({
    author: r.authorAttribution?.displayName || 'Google user',
    rating: r.rating,
    text: (r.text?.text || r.originalText?.text || '').trim(),
    time: r.relativePublishTimeDescription || '',
    source: 'google',
  }));
  const next = { google: { place_id: PLACE_ID, rating: d.rating, total: d.userRatingCount, reviews }, manual: current.manual || [] };
  writeFileSync(target, JSON.stringify(next, null, 2) + '\n');
  console.log(`[reviews] refreshed: ${d.rating} stars, ${d.userRatingCount} reviews, ${reviews.length} recent`);
} catch (e) {
  console.log('[reviews] refresh failed, keeping committed reviews.json:', e.message);
}
